### PR TITLE
Vehicle boarding rework

### PR DESCRIPTION
**Pull request by nocico**
An NPC outside a vehicle with in_vehicle set and the resulting debug message spam prompted me to make this pull request. This:

1) gets rid of state such as in_vehicle, passenger_id and replaces them with computation. un/board_vehicle remains responsible for the effects of moving within a vehicle (like skidding), but in the future it would simplify the code a whole lot if they were deleted and the effects instead hooked in from setpos or something like that.

2) as a byproduct allows monsters to board vehicles and move with them. However updating the code to be fully monster+vehicle aware is TBD.

### CODE REVIEW

File: src/vehicle.cpp [hunk 1]
```diff
@@ -2787,27 +2784,33 @@ std::vector<int> vehicle::boarded_parts() const
...
     p = part_with_feature (p, VPFLAG_BOARDABLE, false);
-    if (p >= 0 && parts[p].has_flag(vehicle_part::passenger_flag))
+    if (p >= 0)
     {
-     const int player_id = parts[p].passenger_id;
-     if( player_id == g->u.getID()) {
-      return &g->u;
-     }
-        return g->npc_by_id( player_id );
+	T *psg = g->critter_at<T> (global_part_pos3(p));
```
Comment (Coolthulhu) [inline src/vehicle.cpp]: This is relatively costly and will get called a lot. If you want it be done this way, make sure it works reasonably fast in rarer edge cases:
* 20+ cars in reality bubble
* A main battle tank (from the tanks mod) sized vehicle
* 3+ NPCs boarding the above vehicle
* 50+ zombies in reality bubble
* All of the above at once

The problem is, your approach can't be easily optimized. Any slowdown here will surely stay with us.

File: src/map.cpp [hunk 2]
```diff
@@ -1320,34 +1318,33 @@ vehicle *map::displace_vehicle( tripoint &p, const tripoint &dp )
 
     // record every passenger inside
     std::vector<int> psg_parts = veh->boarded_parts();
-    std::vector<player *> psgs;
+    std::vector<Creature *> psgs;
+    const tripoint old_veh_pos = veh->global_pos3();
     for( auto &prt : psg_parts ) {
+        auto psg = veh->get_passenger( prt );
+        const tripoint part_pos = old_veh_pos + veh->parts[prt].precalc[0];
         psgs.push_back( veh->get_passenger( prt ) );
```
Comment (BevapDin) [inline src/map.cpp]: Why is this calling `get_passenger` again, you already know who is there (it's `psg`).

File: src/map.cpp [hunk 3]
```diff
@@ -1320,34 +1318,33 @@ vehicle *map::displace_vehicle( tripoint &p, const tripoint &dp )
...
+        if( psg->pos() != part_pos ) {
+            debugmsg( "Passenger/part position mismatch: passenger %d,%d,%d, part %d %d,%d,%d",
+                g->u.posx(), g->u.posy(), g->u.posz(),
+                prt,
+                part_pos.x, part_pos.y, part_pos.z );
+            continue;
```
Comment (BevapDin) [inline src/map.cpp]: Why continue here, the loop body ends afterwards anyway?

File: src/vehicle.cpp [hunk 4]
```diff
@@ -2787,27 +2784,33 @@ std::vector<int> vehicle::boarded_parts() const
...
+	if (psg && psg->in_vehicle()) {
```
Comment (BevapDin) [inline src/vehicle.cpp]: This `in_vehicle` checks looks a bit redundant. All it does is: ensure there is a vehicle with a part that has the boardable flag. But that has already been done in this function. So what's the point? In other words: when will this ever return `false`?

File: src/vehicle.cpp [hunk 5]
```diff
@@ -6460,9 +6463,8 @@ void vehicle::calc_mass_center( bool use_precalc ) const
             m_part += j.weight();
         }
 
-        if( pi.has_flag( VPFLAG_BOARDABLE ) && parts[i].has_flag( vehicle_part::passenger_flag ) ) {
-            const player *p = get_passenger( i );
-            // Sometimes flag is wrongly set, don't crash!
+        if( pi.has_flag( VPFLAG_BOARDABLE ) ) {
+            const Creature *p = get_passenger( i );
```
Comment (BevapDin) [inline src/vehicle.cpp]: Btw. (not really needed in this PR): checking the boardable flag *before* calling `get_passenger` is redundant as well. Just call the function, if the target part does not have the flag, the function will return `nullptr` anyway.

File: src/creature.h [hunk 6]
```diff
@@ -296,6 +296,7 @@ class Creature
         virtual const tripoint &pos() const = 0;
 
         virtual void setpos( const tripoint &pos ) = 0;
+        virtual bool in_vehicle() const;
```
Comment (BevapDin) [inline src/creature.h]: Why is it virtual?

File: src/map.cpp [hunk 7]
```diff
@@ -498,7 +498,7 @@ bool map::vehact( vehicle &veh )
         for( int boarded : veh.boarded_parts() ) {
             if( veh.part_with_feature( boarded, VPFLAG_CONTROLS, true ) >= 0 ) {
                 controlled = true;
-                player *passenger = veh.get_passenger( boarded );
+                player *passenger = dynamic_cast<player *>(veh.get_passenger( boarded ));
```
Comment (BevapDin) [inline src/map.cpp]: The whole point of the template parameter to `game::critter_at` (and in turn to `vehicle::get_passenger`) is so callers *don't* have to manually cast the returned pointer. Instead the function is doing it. Just call it as `get_passenger<player>( ... )` and get rid of the cast.

**Comment by nocico** [comment]
in_vehicle() is prematurely virtual and called redundantly because I thought there might be monsters that could be on the vehicle tile yet not in the vehicle. It could be ripped out if you prefer.

All the other things are down to copy-paste etc. and should be fixed :D 

I'm not too worried about performance except for the loop over all active npcs in critter_at, which really could be a problem but with effort could be optimized if it is. Without really checking everything else seems to be pretty bounded already :D Is there an easy way to benchmark the game?

**Pull request closed without merging**